### PR TITLE
torch.xlogy: Use wrapped_scalar_tensor / gpu_with_scalars to speed up GPU kernel.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49926 torch.xlogy: Use wrapped_scalar_tensor / gpu_with_scalars to speed up GPU kernel.**

While investigating https://github.com/pytorch/pytorch/issues/49758, I changed the xlogy kernel to use the recommended wrapped_scaler_tensor pattern instead of moving the scalar to the GPU as a tensor.
While this doesn't avoid a synchronization (there is no synchronization in the move, as its done via fill), this does significantly speed up the GPU kernel (almost ~50%, benchmark in PR comments).

From looking at the nvprof output, it looks like this code path avoids broadcasting.  Aside: this seems unnecessary, as there is nothing special from the point-of-view of broadcasting whether the Tensor
is ()-sized or marked as a wrapped_scalar.  Still, this is a useful change to make as we avoid extra kernel launches and dispatches to create and fill the tensor.

Differential Revision: [D25724215](https://our.internmc.facebook.com/intern/diff/D25724215)

### CODE REVIEW

**Comment by gchanan** [comment]
Benchmark script:
```
import torch
import torch.utils.benchmark as benchmark
from itertools import product

results = []
sizes = [1, 64, 1024, 10000]

for b, n in product(sizes, sizes):
    label = 'xlogy'
    sub_label = f'[{b}, {n}]'
    x_cpu = torch.randn([b, n], device='cpu')
    x_cuda = torch.randn([b, n], device='cuda')
    results.append(benchmark.Timer(
        stmt='torch.xlogy(x_cuda, 5)',
        setup='',
        globals={'x_cuda': x_cuda},
        label=label+'_cuda',
        sub_label=sub_label,
        description='time',
    ).blocked_autorange(min_run_time=1))
    results.append(benchmark.Timer(
        stmt='torch.xlogy(x_cpu, 5)',
        setup='',
        globals={'x_cpu': x_cpu},
        label=label+'_cpu',
        sub_label=sub_label,
        description='time',
    ).blocked_autorange(min_run_time=1))

compare = benchmark.Compare(results)
compare.print()
```

**Comment by gchanan** [comment]
Benchmark results:

Before:
```
[--------- xlogy_cuda --------]
                      |   time
1 threads: --------------------
      [1, 1]          |    19.8
      [1, 64]         |    22.5
      [1, 1024]       |    21.8
      [1, 10000]      |    21.5
      [64, 1]         |    20.8
      [64, 64]        |    21.4
      [64, 1024]      |    20.9
      [64, 10000]     |    48.1
      [1024, 1]       |    20.9
      [1024, 64]      |    21.3
      [1024, 1024]    |    73.9
      [1024, 10000]   |   652.8
      [10000, 1]      |    20.3
      [10000, 64]     |    48.1
      [10000, 1024]   |   653.2
      [10000, 10000]  |  6312.7

Times are in microseconds (us).

[---------- xlogy_cpu -----------]
                      |     time
1 threads: -----------------------
      [1, 1]          |        6.0
      [1, 64]         |        6.7
      [1, 1024]       |       19.3
      [1, 10000]      |      132.4
      [64, 1]         |        7.1
      [64, 64]        |       56.6
      [64, 1024]      |      815.8
      [64, 10000]     |     7842.3
      [1024, 1]       |       18.6
      [1024, 64]      |      805.9
      [1024, 1024]    |    12683.8
      [1024, 10000]   |   142487.5
      [10000, 1]      |      126.9
      [10000, 64]     |     8103.1
      [10000, 1024]   |   144214.1
      [10000, 10000]  |  1405848.1
```

After:
```
[--------- xlogy_cuda --------]
                      |   time
1 threads: --------------------
      [1, 1]          |    13.8
      [1, 64]         |    13.8
      [1, 1024]       |    14.2
      [1, 10000]      |    14.1
      [64, 1]         |    14.1
      [64, 64]        |    14.2
      [64, 1024]      |    14.7
      [64, 10000]     |    23.2
      [1024, 1]       |    19.7
      [1024, 64]      |    19.4
      [1024, 1024]    |    39.8
      [1024, 10000]   |   374.1
      [10000, 1]      |    13.9
      [10000, 64]     |    23.1
      [10000, 1024]   |   373.8
      [10000, 10000]  |  3644.9

Times are in microseconds (us).

[---------- xlogy_cpu -----------]
                      |     time
1 threads: -----------------------
      [1, 1]          |        5.6
      [1, 64]         |        7.3
      [1, 1024]       |       18.5
      [1, 10000]      |      129.0
      [64, 1]         |        6.5
      [64, 64]        |       57.7
      [64, 1024]      |      875.4
      [64, 10000]     |     7719.4
      [1024, 1]       |       21.2
      [1024, 64]      |      800.5
      [1024, 1024]    |    12640.4
      [1024, 10000]   |   144726.7
      [10000, 1]      |      127.5
      [10000, 64]     |     7626.6
      [10000, 1024]   |   141238.0
      [10000, 10000]  |  1367694.0
```

**Comment by gchanan** [comment]
Wait, doesn't this also change the type promotion behavior (in a good way?), which would also affect the performance:

Before:
```
>>> torch.xlogy(torch.tensor(5), 5.).dtype
torch.float64
```

After:
```
>>> torch.xlogy(torch.tensor(5), 5.).dtype
torch.float32
```

Which matches the `torch.add` behavior (both Before and After):
```
torch.add(torch.tensor(5), 5.).dtype
torch.float32
```

**Comment by gchanan** [comment]
Note that the benchmark results are valid because type promotion didn't change for that case.

**Comment by mruberry** [comment]
Per offline conversation: add a test in test_binary_ufuncs.py.

**Comment by gchanan** [comment]
As pointed out by @mruberry, `torch.xlogy` wasn't released in 1.7, so I'm removing the bc-breaking label.

**Review by mruberry — APPROVED** [review]
Cool!
